vault/resource_managed_keys.go

```diff
@@ -0,0 +1,661 @@
+	// This also handles the case of an update on 'name'
+	for k := range oldKeySet {
+		if !newKeySet[k] {
+			// Delete single key type
+			if diags := deleteSingleManagedKey(d, meta, keyType, k); diags != nil {
+				return diags
+			}
+		}
+	}
+
+	return nil
+}
+
+func createUpdateManagedKeys(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	if _, ok := d.GetOk(consts.FieldAWS); ok {
+		if diags := writeManagedKeysData(d, meta, consts.FieldAWS); diags != nil {
+			return diags
+		}
+	}
+
+	if _, ok := d.GetOk(consts.FieldPKCS); ok {
+		if diags := writeManagedKeysData(d, meta, consts.FieldPKCS); diags != nil {
+			return diags
+		}
+	}
+
+	if _, ok := d.GetOk(consts.FieldAzure); ok {
+		if diags := writeManagedKeysData(d, meta, consts.FieldAzure); diags != nil {
+			return diags
+		}
+	}
+
+	// set ID to 'default'
+	d.SetId("default")
+
+	return readManagedKeys(ctx, d, meta)
+}
+
+func getRedactedFields(d *schema.ResourceData, providerType, name string,
+	fields []string, m map[string]interface{},
+) map[string]interface{} {
+	prefix := fmt.Sprintf("%s.%d.", providerType, getHashFromName(name))
+	for _, field := range fields {
+		k := prefix + field
+		if v, ok := d.GetOk(k); ok {
+			m[field] = v
+		}
+	}
+
+	return m
+}
+
+func readAndSetManagedKeys(d *schema.ResourceData, meta interface{},
+	providerType string, sm map[string]string, redactedFields []string) error {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return e
+	}
+
+	var keyType string
+	switch providerType {
+	case consts.FieldAWS:
+		keyType = kmsTypeAWS
+
+	case consts.FieldPKCS:
+		keyType = kmsTypePKCS
+
+	case consts.FieldAzure:
+		keyType = kmsTypeAzure
+
+	default:
+		return fmt.Errorf("received unexpected provider type %s", providerType)
+	}
+
+	p := fmt.Sprintf("%s/%s", "sys/managed-keys", keyType)
+	log.Printf("[DEBUG] Listing data from Vault at %s", p)
+	resp, err := client.Logical().List(p)
+	if err != nil {
+		return err
+	}
+
+	if resp == nil {
+		return nil
+	}
+
+	handlers := map[string]func() schemaMap{
+		kmsTypeAWS:   managedKeysAWSConfigSchema,
+		kmsTypePKCS:  managedKeysPKCSConfigSchema,
+		kmsTypeAzure: managedKeysAzureConfigSchema,
+	}
+
+	var data []interface{}
+	if v, ok := resp.Data["keys"]; ok {
+		for _, name := range v.([]interface{}) {
+			m := make(map[string]interface{})
+			path := getManagedKeysPath(keyType, name.(string))
+			log.Printf("[DEBUG] Reading from Vault at %s", path)
+			resp, err := client.Logical().Read(path)
+			if err != nil {
+				return err
+			}
+
+			if resp == nil {
+				continue
+			}
+
+			for k := range handlers[keyType]() {
+				// Map TF schema fields to Vault API
+				vaultKey := k
+				if v, ok := sm[k]; ok {
+					vaultKey = v
+				}
+
+				if v, ok := resp.Data[vaultKey]; ok {
+					m[k] = v
+				}
+			}
+
+			// get these from TF config since they are
+			// returned as "redacted" from Vault
+			m = getRedactedFields(d, providerType, name.(string), redactedFields, m)
```

Comment:
nit: since `getRedactedFields()` is mutating `m` , a better name might be `updateRedactedFields` instead. Not sure it needs to return the map it just mutated either.

Reply:
Renamed in [2347120](https://github.com/hashicorp/terraform-provider-vault/pull/1508/commits/2347120bc674a335eb4ca5a02a893c098bd2cb4e)